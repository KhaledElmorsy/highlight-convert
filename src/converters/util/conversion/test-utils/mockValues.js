export default [
  'usd',
  'egp',
  'eur',
  'gbp',
  'cad',
  'jpy',
  'aed',
].map((id, amount) => ({amount, unit: {id}}))
