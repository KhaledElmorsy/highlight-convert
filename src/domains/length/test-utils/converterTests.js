/**
 * Data set up 
 * {@link https://docs.google.com/spreadsheets/d/1Ml0Oc5__v8jOMlOIk5Wa-p6mJOhWEBeUZPC5KPf2AYk/edit?usp=sharing here}
 * (Google Docs)
 */
export default [
  [
    { ft: 6 },
    {
      m: 1.82871076,
      cm: 182.87107589,
      mm: 1828.71075891,
      km: 0.00182871,
      in: 71.99634258,
      ft: 6,
      yd: 1.99989637,
      mi: 0.00113631,
      thou: 71996.52544956,
    },
  ],
  [
    { cm: 200 },
    {
      m: 2,
      cm: 200,
      mm: 2000,
      km: 0.002,
      in: 78.74,
      ft: 6.562,
      yd: 2.1872,
      mi: 0.0012427,
      thou: 78740,
    },
  ],
  [
    { yd: 5 },
    {
      m: 4.572,
      cm: 457.2,
      mm: 4572,
      km: 0.004572,
      in: 180,
      ft: 15.001,
      yd: 5,
      mi: 0.0028409,
      thou: 180000,
    },
  ],
];
