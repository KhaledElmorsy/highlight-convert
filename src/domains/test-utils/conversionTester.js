export default function conversionTester(converter, unitMap) {
  return async (input, expected) => {
    const inputVector = Object.entries(input).reduce(
      (_, [id, amount]) => ({
        amount,
        unit: unitMap[id],
      }),
      null
    );

    const conversions = await converter.convert(inputVector);
    conversions.forEach(({ amount, unit }) => {
      expect(amount).toBe(expected[unit.id]);
    });
  };
}
