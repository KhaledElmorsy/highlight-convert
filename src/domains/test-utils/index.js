export { default as conversionTester } from './conversionTester';
export { default as matchTester } from './matchTester';
export { default as validateRenderSettings } from './validateRenderSettings';
