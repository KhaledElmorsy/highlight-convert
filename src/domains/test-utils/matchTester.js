/**
 * Test a unit matching method.
 * @param {(string: string) => Promise<Match[]>} match 
 * @param {{[id: Unit['id']]: Unit}} unitMap 
 * @returns 
 */
export default function matchTester(match, unitMap) {
  /**
   * @param {string} string
   * @param {{[id: Unit['id']]: number}[]} expected
   */
  return async (string, expectedValues) => {
    expect(await match(string)).toEqual(
      expect.arrayContaining(
        expectedValues.map((value) => {
          const [[id, amount]] = Object.entries(value);
          return expect.objectContaining({
            value: expect.objectContaining({
              unit: unitMap[id],
              amount,
            }),
          });
        })
      )
    );
  };
}
