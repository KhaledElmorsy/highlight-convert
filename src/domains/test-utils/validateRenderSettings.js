/**
 * Assert that all unit IDs in a render settings object belong to units in a
 * specified set.
 * @param {ConversionRenderSettings<Unit>} renderSettings
 * @param {{[id: string]: Unit}} unitMap
 */
export default function validateRenderSettings(renderSettings, unitMap) {
  const checkID = (id) => expect(Object.hasOwn(unitMap, id)).toBeTruthy();
  for (let [key, value] of Object.entries(renderSettings)) {
    switch (key) {
      case 'mainUnitID':
      case 'secondaryUnitID':
        checkID(value);
        break;
      case 'featuredUnitIDs':
        value.forEach(checkID);
        break;
      case 'groups':
        value.forEach((group) => {
          group.unitIDs.forEach(checkID);
        });
        break;
      case 'unitTemplates':
        Object.keys(value).forEach(checkID);
      // no-default
    }
  }
}
