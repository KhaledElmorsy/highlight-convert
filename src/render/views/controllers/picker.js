import Select from '@ui5/webcomponents/dist/Select';

/** @typedef {import('@ui5/webcomponents/dist/Option').default} Option */

/**
 * @template Value
 * @typedef {import('./typedefs').ControllerViewExtension<Value>} ControllerViewExtension
 */

/**
 * @typedef {object} PickerSettings
 * @prop {(option: any) => any} [mapOptions]
 * @prop {object} [keys] If options are objects, specify properties to render
 * as the main and secondary columns
 * @prop {string} [keys.main] Default: `name`
 * @prop {string|null} [keys.sub=null] Optional second column. Ignored if `null` (default).
 */

/**
 * Create a basic picker.
 *
 * Options can be strings or objects. When using objects, specifiy the property
 * containing the main text to render, and optionally, a second property to render
 * in a second column.
 * @template {string | {[key: string]: any}} Option
 * @param {object} args
 * @param {Option[]} args.options
 * @param {Option} args.value
 * @param {PickerSettings} [args.settings]
 * @param {(value:  Option) => void} [args.onChange]
 * @returns {Select & ControllerViewExtension<Option>}
 */
export default function picker({
  options,
  value,
  settings: {
    keys: { main = 'name', sub = null } = {},
    mapOptions = (options) => options,
  } = {},
  onChange = () => {},
}) {
  /** @type {Select} */
  const selectElement = document.createElement('ui5-select');
  const areObjects = typeof options[0] === 'object';

  function getOptionElements(selectedValue) {
    const optionElements = options.map(mapOptions).map((option, i) => {
      const [mainText, subText] = areObjects
        ? [option[main], option[sub] ?? null]
        : [option, null];

      /** @type {Option} */
      const element = document.createElement('ui5-option');
      element.textContent = mainText;
      element.setAttribute('data-index', i); // To cross-reference he relevant option

      // Compare the (pre-map) option with the passed picked 'value' to set the selected option
      const initalOption = options[i];
      if (JSON.stringify(initalOption) === JSON.stringify(selectedValue)) {
        element.setAttribute('selected', '');
      }

      if (subText !== null) element.setAttribute('additional-text', subText);
      return element;
    });

    return optionElements;
  }

  selectElement.append(...getOptionElements(value));
  selectElement.addEventListener('change', ({ detail: { selectedOption } }) => {
    const index = selectedOption.getAttribute('data-index');
    onChange(options[index]);
  });

  /** @type {ControllerViewExtension<Option>} */
  const controllerExtension = {
    acceptVisitor(viewVisitor) {
      viewVisitor.picker({
        options,
        keys: areObjects ? { main, sub } : null,
        mapOptions,
        element: selectElement,
      });
    },
    setValue(newValue) {
      const tempParent = document.createElement('div');
      tempParent.append(...getOptionElements(newValue));
      this.innerHTML = tempParent.innerHTML;
    },
  };

  Object.assign(selectElement, controllerExtension);

  return selectElement;
}
