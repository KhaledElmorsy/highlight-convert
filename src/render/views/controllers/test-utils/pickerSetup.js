// Prepped setup for views that pick from different passed options

const options = {
  string: ['testA', 'testB', 'testC'],
  object: [
    { id: 'a', name: 'Integer-test', type: 'number' },
    { id: 'b', name: 'Float-test', type: 'number' },
    { id: 'c', name: 'String-test', type: 'string' },
  ],
};

export const picker = {
  string: {
    options: options.string,
    mapOptions: (options) => options.toUpperCase(),
  },
  object: {
    options: options.object,
    mapOptions: ({ id, name, type }) => ({
      title: name.toUpperCase(),
      subtitle: id.toUpperCase(),
    }),
    keys: { main: 'title', sub: 'subtitle' },
  },
};

export const comboBox = {
  string: {
    options: options.string,
    mapOptions: (options) => options.toUpperCase(),
  },
  object: {
    options: options.object,
    mapOptions: ({ id, name, type }) => ({
      title: name.toUpperCase(),
      subtitle: id.toUpperCase(),
      group: type.toUpperCase(),
    }),
    keys: { main: 'title', sub: 'subtitle', group: 'group' },
  },
};
