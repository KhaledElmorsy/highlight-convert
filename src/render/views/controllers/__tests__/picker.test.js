import picker from '../picker';
import { picker as setup } from '../test-utils/pickerSetup';

jest.mock('@ui5/webcomponents/dist/Select', () => {});

describe.each([
  [
    'String',
    {
      init: {
        options: setup.string.options,
        value: setup.string.options[1],
        settings: {
          mapOptions: setup.string.mapOptions,
        },
      },
      newValue: setup.string.options[0],
      visitorArgs: {
        options: setup.string.options,
        mapOptions: setup.string.mapOptions,
        keys: null,
      },
    },
  ],
  [
    'Object',
    {
      init: {
        options: setup.object.options,
        value: setup.object.options[1],
        settings: {
          mapOptions: setup.object.mapOptions,
          keys: setup.object.keys,
        },
      },
      newValue: setup.object.options[0],
      visitorArgs: {
        options: setup.object.options,
        mapOptions: setup.object.mapOptions,
        keys: setup.object.keys,
      },
    },
  ],
])('%s options', (_, { init, newValue, visitorArgs }) => {
  const testPicker = picker(init);
  it('Renders correctly', () => {
    expect(testPicker).toMatchSnapshot();
  });
  it('Programatically updates values', () => {
    testPicker.setValue(newValue);
    expect(testPicker).toMatchSnapshot();
  });
  it('Dispatches the correct visitor method with relevant arguments', () => {
    const visitor = {
      picker: jest.fn(),
    };
    testPicker.acceptVisitor(visitor);
    expect(visitor.picker).toHaveBeenCalledWith({
      ...visitorArgs,
      element: testPicker,
    });
  });
});
