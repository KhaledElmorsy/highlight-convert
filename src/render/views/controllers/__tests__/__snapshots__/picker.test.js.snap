// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object options Programatically updates values 1`] = `
<ui5-select>
  <ui5-option
    additional-text="A"
    data-index="0"
    selected=""
  >
    INTEGER-TEST
  </ui5-option>
  <ui5-option
    additional-text="B"
    data-index="1"
  >
    FLOAT-TEST
  </ui5-option>
  <ui5-option
    additional-text="C"
    data-index="2"
  >
    STRING-TEST
  </ui5-option>
</ui5-select>
`;

exports[`Object options Renders correctly 1`] = `
<ui5-select>
  <ui5-option
    additional-text="A"
    data-index="0"
  >
    INTEGER-TEST
  </ui5-option>
  <ui5-option
    additional-text="B"
    data-index="1"
    selected=""
  >
    FLOAT-TEST
  </ui5-option>
  <ui5-option
    additional-text="C"
    data-index="2"
  >
    STRING-TEST
  </ui5-option>
</ui5-select>
`;

exports[`String options Programatically updates values 1`] = `
<ui5-select>
  <ui5-option
    data-index="0"
    selected=""
  >
    TESTA
  </ui5-option>
  <ui5-option
    data-index="1"
  >
    TESTB
  </ui5-option>
  <ui5-option
    data-index="2"
  >
    TESTC
  </ui5-option>
</ui5-select>
`;

exports[`String options Renders correctly 1`] = `
<ui5-select>
  <ui5-option
    data-index="0"
  >
    TESTA
  </ui5-option>
  <ui5-option
    data-index="1"
    selected=""
  >
    TESTB
  </ui5-option>
  <ui5-option
    data-index="2"
  >
    TESTC
  </ui5-option>
</ui5-select>
`;
